src/dstore/pmix_esh.c

```diff
@@ -228,22 +491,64 @@ int _esh_init(pmix_info_t info[], size_t ninfo)
         return PMIX_SUCCESS;
     }
     else {
+        int rc = 0;
+        int tbl_idx;
+        nspace_tbl_entry_t *nspace_tbl_entry = NULL;
+
+        if (NULL != info) {
+            for (n=0; n < ninfo; n++) {
+                if (0 == strcmp(PMIX_USERID, info[n].key)) {
+                    _jobuid = info[n].value.data.uint32;
+                    _setjobuid = 1;
+                    continue;
+                }
+            }
+        }
+
+        rc = _esh_nspace_tbl_add(&tbl_idx);
+        if (PMIX_SUCCESS != rc) {
+            PMIX_ERROR_LOG(rc);
+            return rc;
+        }
+
+        rc = _esh_nspace_tbl_map_idx(pmix_globals.myid.nspace, tbl_idx);
+        if (PMIX_SUCCESS != rc) {
+            PMIX_ERROR_LOG(rc);
+            return rc;
+        }
+
+        nspace_tbl_entry = (nspace_tbl_entry_t *)(&_nspace_tbl[tbl_idx]);
+        if (NULL == nspace_tbl_entry) {
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            return PMIX_ERROR;
+        }
+
+        nspace_tbl_entry->jobuid = _jobuid;
+
+        if (0 > asprintf(&nspace_tbl_entry->nspace_path, "%s/%u", _base_path, _jobuid)) {
```

Comment:
Server has to pass the full path to the namespace dir. you don't need to assemble it here and (as I've said before) care about jobuid. 
`PMIX_DSTORE_ESH_BASE_PATH` must contain the full path to this user's jobuid base directory.
So the rule is following: "Client shouldn't contain any path assembly logic"
Here the server might want to separate several sessions of the same user using directories. If we keep all path building logic on the server and provide only the final path to client - we will have to do less changes.